Blog on mpOpenAPI (#4353)
OpenLiberty/blogs: posts/2025-03-28-Building-Better-APIs:-Documenting-APIs-in-Open-Liberty-using-mpOpenAPI.adoc

blog_description: "This blog explores the OpenAPI Specification and its integration with Open Liberty using MicroProfile OpenAPI. It covers the importance of API documentation, the role of OpenAPI and Swagger, and a step-by-step guide to using MicroProfile OpenAPI in an Open Liberty project. Readers will learn about various MicroProfile OpenAPI annotations with practical examples and see how they are rendered in Swagger UI."
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
additional_authors: 
- name: Abhinav K
  github: https://github.com/itsabhinavk
  image: https://avatars0.githubusercontent.com/itsabhinavk
---
= Building Better APIs: Documenting APIs in Open Liberty using MicroProfile OpenAPI
Sweetty P Devassy <https://github.com/sweettypdevassy>
:imagesdir: /
:url-prefix:
:url-about: /

> With evolving technology, we are building a huge number of highly complicated APIs. Documenting these APIs requires effort and is often seen as a tedious task. Furthermore, complex APIs make documentation even more challenging. In reality, API documentation serves as a means of sharing up-to-date information between development teams. This crucial aspect later led to the development of Swagger and later OpenAPI specification. So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with MicroProfile OpenAPI features.

Reviewer: ```suggestion
As technology has evolved and our applications have become more complicated, so too have our APIs. We are now building a huge number of highly complicated APIs. Documenting them requires effort and is often seen as a tedious task. Furthermore, the complex nature of these APIs can make documentation even more challenging. But why does documentation of our APIs matter? API documentation serves as a means of sharing up-to-date information between development teams, acting as a sort of how-to manual for developers that are looking to build things with your API. But having a standardized way of documenting our APIs so that we can all easily understand other APIs and what they do was a challenge... at least until the development of [Swagger](https://swagger.io/) and later the [OpenAPI](https://www.openapis.org/) specification. Having a structured approach to documentation helps developers improve collaboration, reduce errors, and accelerate development cycles.

So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with [MicroProfile OpenAPI](https://microprofile.io/specifications/open-api/) features to easily and effectively document your APIs.
```

== What is OpenAPI?  

*OpenAPI* is a *standardised specification* for describing *RESTful APIs* in a machine-readable format (JSON or YAML). It allows developers to *define, document, and share API structures* in a way that is both human-readable and programmatically useful. At its core, an OpenAPI definition includes:  

* *Info* – API title, version, description  

* *Servers* – Base URLs of the API  

* *Paths* – Endpoints (/users, /orders)  

* *Methods* – GET, POST, PUT, DELETE  

* *Parameters* – Query, path, header, or body inputs  

* *Responses* – Expected status codes and return data  

* *Security* – Authentication methods (OAuth, API keys, etc.)  

> Swagger UI, developed by SmartBear Software, allows developers to visualize API documentation and test APIs interactively. 

Reviewer: ```suggestion
The OpenAPI specification can be used together with [Swagger UI](https://swagger.io/tools/swagger-ui/) to allow developers to visualize the APIs and test them interactively. The great part about this integration is that the UI is automatically generated from your OpenAPI specification.
```

== MicroProfile OpenAPI and Open Liberty  

> For *Open Liberty developers, MicroProfile OpenAPI* library provides the integration to add the OpenAPI specification to Open Liberty. Integrating *MicroProfile OpenAPI* is crucial as it *automatically generates OpenAPI-compliant documentation* for JAX-RS services, eliminating the need for additional manual effort. Developers are now able to maintain *consistent, accurate, and up-to-date API documentation*. Additionally, combining *MicroProfile OpenAPI with Swagger UI* enhances the developer and user experience by providing *interactive API documentation*, allowing users to visualise endpoints, test API calls, and understand request/response structures in a more intuitive way. This removes the need for manually written API documentation, reducing maintenance overhead and minimising the risk of outdated or inconsistent documentation. *MicroProfile OpenAPI streamlines API development* in Open Liberty, making it easier for teams to build, expose, and manage RESTful services efficiently.  

Reviewer: ```suggestion
For Open Liberty developers, [MicroProfile OpenAPI](https://microprofile.io/specifications/open-api/) library can be used to integrate and make use of the OpenAPI specification. As well as enabling the documentation standards offered through the OpenAPI specification, the MicroProfile OpenAPI also eliminates some of the need for additional manual effort by offering the ability to automatically generate OpenAPI-compliant documentation for JAX-RS services. Saving even more time and effort! Making use of MicroProfile OpenAPI, Open Liberty developers are able to maintain consistent, accurate, and up-to-date API documentation. 

Additionally, MicroProfile OpenAPI can be used seamlessly with the Swagger UI to enhance the developer and user experience by providing interactive API documentation, allowing users to visualise endpoints, test API calls, and understand request/response structures in a more intuitive way. This removes the need for manually written API documentation, reducing maintenance overhead and minimising the risk of outdated or inconsistent documentation. 

In summary, MicroProfile OpenAPI streamlines API development in Open Liberty, making it easier for teams to build, expose, and manage RESTful services efficiently.  
```

The MicroProfile OpenAPI feature is responsible for generating this documentation. This eliminates the need for external configuration files and ensures the documentation evolves alongside the codebase.

Ultimately, MicroProfile OpenAPI annotations play a vital role in creating useful, up-to-date documentation. Because they reside within the source code, they make it easier for teams to maintain alignment between implementation and documentation—benefiting both developers and end users who need to understand how to interact with the API effectively.

== Influence  

> As *API architectures evolve,* OpenAPI, the industry-standard specification for *documenting REST APIs,* is influencing how *GraphQL* and *gRPC* APIs handle documentation and interoperability. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability. 

Reviewer: ```suggestion
As API architectures evolve, OpenAPI, the industry-standard specification for documenting REST APIs, is influencing how other tools and specification's APIs handle documentation and interoperability. Among those being influenced is [GraphQL](https://graphql.org/learn/) and [gRPC](https://grpc.io/). GraphQL is a query language for your API, and a server-side runtime for executing queries using a type system you define for your data. gRPC is a modern open source high performance Remote Procedure Call (RPC) framework that can run in any environment. It can efficiently connect services in and across data centers with pluggable support for load balancing, tracing, health checking and authentication. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability. 

For example:
```

> * OpenAPI’s structured *JSON/YAML format* has influenced *GraphQL’s SDL (Schema Definition Language),* where types, queries, and mutations are explicitly defined. 

Reviewer: ```suggestion
* OpenAPI’s structured JSON/YAML format has influenced GraphQL’s SDL (Schema Definition Language), where types, queries, and mutations are explicitly defined. 
```

> * Tools like *Swagger UI*, which originated alongside the Swagger project and now commonly accompany OpenAPI specifications, have inspired similar tools in other ecosystems. For example, *gRPC UI* offers an interactive interface to explore and test gRPC methods, much like Swagger UI does for REST APIs—though both Swagger UI and gRPC UI are independent, community-driven tools.  

Reviewer: ```suggestion
* Tools like Swagger UI, which originated alongside the Swagger project and now commonly accompany OpenAPI specifications, have inspired similar tools in other ecosystems. For example, gRPC UI offers an interactive interface to explore and test gRPC methods, much like Swagger UI does for REST APIs—though both Swagger UI and gRPC UI are independent, community-driven tools.  
```

== Conclusion  

> In today's world of *microservices and cloud computing,* standardised API documentation is more important than ever. OpenAPI has become the foundation for creating *consistent, accessible, and well-documented APIs,* ensuring seamless integration across diverse systems. As APIs continue to evolve, having a structured approach to documentation helps developers *improve collaboration, reduce errors, and accelerate development cycles.* For those working with *Open Liberty,* exploring *MicroProfile OpenAPI* offers a way to generate OpenAPI-compliant documentation, making API management more efficient. As we embrace automation and interoperability in modern software development, OpenAPI will have an important role to play.

Reviewer: ```suggestion
In today's world of microservices and cloud computing, standardised API documentation is more important than ever. OpenAPI has become the foundation for creating consistent, accessible, and well-documented APIs, ensuring seamless integration across diverse systems. As APIs continue to evolve, having a structured approach to documentation helps developers improve collaboration, reduce errors, and accelerate development cycles. For those working with Open Liberty, exploring MicroProfile OpenAPI offers a way to generate OpenAPI-compliant documentation, making API management more efficient. As we embrace automation and interoperability in modern software development, OpenAPI will have an important role to play.
```

> OpenLiberty.io site links:

Reviewer: ```suggestion
Want to get hands-on with this and try it out in our own Open Liberty applications?
Either follow along with our interactive guide, or, check out the documentation we have available too:
```

> link:https://openliberty.io/guides/microprofile-openapi.html[Documenting RESTful APIs]

Reviewer: ```suggestion
* link:https://openliberty.io/guides/microprofile-openapi.html[Guide: Documenting RESTful APIs]
```

> link:https://openliberty.io/docs/25.0.0.1/documentation-openapi.html[API documentation with OpenAPI]

Reviewer: ```suggestion
* link:https://openliberty.io/docs/25.0.0.1/documentation-openapi.html[Docs: API documentation with OpenAPI]
```
